### ISSUE TITLE
All Nan Hinv

### ISSUE THREAD

**Issue by Calmepro777** [IST-DASLab/OBC#1]
Hello,

This is such an amazing work and I am trying to have a clear understanding of OBC by runing the code on my machine.

During debugging, when I printed out the Hinv of each layer, I noticed that some Hinvs have all Nan values. I guess that is because the corresponding H is not invertible or have zero or Nan determinant. 

I wonder if you could suggest any potential solutions.

Thank you in advance.

**Comment by Calmepro777** [comment]
In addition, in https://github.com/IST-DASLab/OBC/blob/main/trueobs.py#L49-L51, I guess the purpose is to average over all input samples while it seems that `self.H` is actually reduced once by multiplying with `total_num_samples/(total_num_samples + cur_batch_size)` and the result for current batch is  averaged over all input samples (divided by `total_num_samples`). (`total_num_samples`: self.nsamples; `cur_batch_size: tmp`)

I wonder if the following substitution for `L49-L51` in `trueobs.py` would make sense, which would be more instinctive to me:

``` python
self.nsamples = self.nsamples + tmp
self.H = self.H + 2 / tmp * (inp.matmul(inp.t())).double()
```

**Comment by Godofnothing** [comment]
@Calmepro777 
Hi, concerning the first question you can add a regularization term to prevent this issue.
I've faced similar issue on some models as well. 

One can specify a parameter `rel_damp`, say `1e-2`, and add term proportional to identity matrix to the hessian before running OBS. More precisely, modify `prepare` method of TrueOBS in `trueobs.py` as follows :
```python
damp = self.rel_damp * torch.diag(self.H).mean()
self.H += damp * self.eye(self.H.shape[0])
```

**Comment by Godofnothing** [comment]
@Calmepro777 concerning the latter question. What you propose should produce the equivalent result if it hadn't been for numerics, since OBS is agnostic to the overall scale factor. But the implementation in repo may be more numerically stable.

**Comment by TianL123** [comment]
Hello, I am interested in your work, it is really a very meaningful job. I used it in yolov5s, and my model is a full-precision model model trained on the NWPU-VHR-10 remote sensing dataset, mAP0.5:89.4%. When I used 4W4A, I found that the effect was somewhat unsatisfactory, and the mAP was only more than 20%. It was found to be very sensitive to activation quantification. If I use 4W8A then it works great, 87.9%. Similarly, 8W4A does not work well. At the same time, my 2:4 sparseness effect is not ideal. I don't know why this is happening. Looking forward to your reply.
